cpp/include/raft/interruptible.hpp

```diff
@@ -0,0 +1,246 @@
+    return s;
+  }
+
+  /**
+   * @brief Get a cancellation token for a CPU thread given by its id.
+   *
+   * @return an object that can be used to cancel the GPU work waited on the given CPU thread.
+   */
+  static auto get_token(std::thread::id thread_id) -> std::shared_ptr<interruptible>
+  {
+    std::lock_guard<std::mutex> guard_get(mutex_);
+    // the following constructs an empty shared_ptr if the key does not exist.
+    auto& weak_store  = registry_[thread_id];
+    auto thread_store = weak_store.lock();
+    if (!thread_store) {
+      thread_store.reset(new interruptible(), [thread_id](auto ts) {
+        std::lock_guard<std::mutex> guard_erase(mutex_);
+        registry_.erase(thread_id);
+        delete ts;
+      });
+      std::weak_ptr<interruptible>(thread_store).swap(weak_store);
+      // Since we're still guarded by the mutex, use this opportunity to initialize the global
+      // cancellation stream. The cancellation stream is only used by non-static cancel(), so it's
+      // safe to initialize it here before the first valid thread store is returned to the user.
+      get_cancellation_stream();
+    }
+    return thread_store;
+  }
+
+  /**
+   * @brief Cancel any current or next call to `interruptible::synchronize` performed on the
+   * CPU thread given by the `thread_id`
+   *
+   * Note, this function uses a mutex to safely get a cancellation token that may be shared
+   * among multiple threads. If you plan to use it from a signal handler, consider the non-static
+   * `cancel_no_throw()` instead.
+   *
+   * @param [in] thread_id a CPU thread, in which the work should be interrupted.
+   *
+   * @throw raft::cuda_error if a CUDA error happens during recording the interruption event.
+   */
+  static inline void cancel(std::thread::id thread_id) { get_token(thread_id)->cancel(); }
+
+  /**
+   * @brief Cancel any current or next call to `interruptible::synchronize` performed on the
+   * CPU thread given by this `interruptible` token.
+   *
+   * Note, this function does not involve thread synchronization/locks, so it should be safe
+   * to call from a signal handler.
+   *
+   * @throw raft::cuda_error if a CUDA error happens during recording the interruption event.
+   */
+  inline void cancel() { RAFT_CUDA_TRY(cancel_no_throw()); }
+
+  /**
+   * @brief Cancel any current or next call to `interruptible::synchronize` performed on the
+   * CPU thread given by this `interruptible` token.
+   *
+   * Note, this function does not involve thread synchronization/locks and not throw any
+   * exceptions, so it's safe to call from a signal handler.
+   *
+   * @return CUDA error code redirected from `cudaEventRecord`.
+   */
+  auto cancel_no_throw() noexcept -> cudaError_t
+  {
+    // This method is supposed to be called from another thread;
+    // multiple calls to it just override each other, and that is ok - the cancellation request
+    // will be delivered (at least once).
+    cancelled_ = true;
```

Comment:
Concurrent, non-atomic writes to the same location is undefined behavior. It should be a `std::atomic_flag` and use `test_and_set(memory_order_relaxed)`.